daemon/cmd/hubble.go

```diff
@@ -172,7 +173,19 @@ func (d *Daemon) launchHubble() {
 		}
 	}
 
-	d.hubbleObserver, err = observer.NewLocalServer(payloadParser, logger,
+	namespaceManager := observer.NewNamespaceManager()
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		namespaceManager.Run(d.ctx)
+		wg.Done()
+	}()
+	defer wg.Wait()
```

Comment:
What happen on Hubble initialization failure? As far as I understand, we'll wait indefinitely on this `wg.Wait()` since `d.ctx` is not cancelled.

To me, the usage of `sync.WaitGroup` here indicate that we need something to finish cleanly from `namespaceManager.Run` before returning, but it doesn't seems to be the case. Instead, we want to shutdown the `namespaceManager` when we return. If correct, I would like to suggest

```suggestion
	ctx, cancel := context.WithCancel(d.ctx)
	defer cancel()
	go namespaceManager.Run(ctx)
```

Reply:
Good point on initialization failure, and the approach to do cancellation seems like the right way to do it.

The waitGroup is just best practice from my perspective. In my opinion you should always ensure every go routine returns (even on shutdown). The waitGroup here is to ensure that it does. We just aren't doing a very (IMO) good job of it in the rest of this function currently (and I don't want to rewrite the whole function right now either).

This is to just avoid potential memory leaks. It's hard to know how code might be used in the future. For example, if we changed agent to restart hubble automatically on failure (right now it does not) then if we don't check the go routine returns before returning, then we might have an infinite number of go routines get started,.

Reply:
Taking another look, before this patch, `launchHubble` is [executed into its own goroutine](https://github.com/cilium/cilium/blob/3e805306fc4125fca3398397acdf11a13f1a9600/daemon/cmd/daemon_main.go#L1871). In the happy path the "launch" goroutine setup a goroutine for the "local server" and a corresponding cleanup goroutine, another goroutine for the "remote server" and a corresponding cleanup goroutine, and then returns. So we start the "launch" goroutine, itself starting 4 other goroutines, and then the "launch" goroutine terminates.

My suggestion then doesn't work because when launchHubble returns then the context is cancelled, and the `namespaceManager` shutdown.

In the current version of the patch, the "launch" goroutine will be stuck on the deferred `wg.Wait()` until `d.ctx` is cancelled, cancelling the local `ctx` and shutting down the `namespaceManager` (which will eventually call `wg.Done()` and allow the "launch" goroutine to return from `launchHubble` and terminate).

If correct, although I agree with you wrt best practice on principle, practically I don't see the point of keeping the waitGroup. Technically it only make it so the "launch" goroutine stick around, which have any purpose anymore. As I see it it is one more goroutine that is sleeping or that we might leak.

Reply:
Interesting, I rebuilt and tested locally with this change but didn't actually notice any issues. I guess that means the namespace manager just wasn't going to expire resources, so I didn't catch it. I'll take another look at this.